etcdctl backup tool is broken

I am trying to backup and restore an etcd cluster and it doesn't seem to be working. I am getting a 


Checking the contents of the directory:
```
core@ip-10-0-0-50 ~ $ sudo find /var/lib/etcd2
/var/lib/etcd2
/var/lib/etcd2/member
/var/lib/etcd2/member/snap
/var/lib/etcd2/member/snap/0000000000000002-00000000000186aa.snap
/var/lib/etcd2/member/snap/0000000000000002-0000000000015f99.snap
/var/lib/etcd2/member/snap/0000000000000002-000000000001fbdd.snap
/var/lib/etcd2/member/snap/0000000000000002-000000000001d4cc.snap
/var/lib/etcd2/member/snap/0000000000000002-000000000001adbb.snap
/var/lib/etcd2/member/wal
/var/lib/etcd2/member/wal/0000000000000001-0000000000011eec.wal
/var/lib/etcd2/member/wal/0000000000000000-0000000000000000.wal
```

Checking the output of the backup tool:

```
core@ip-10-0-0-50 ~ $ sudo etcdctl backup --data-dir /var/lib/etcd2 --backup-dir backup2
core@ip-10-0-0-50 ~ $ sudo find backup2
backup2
backup2/member
backup2/member/snap
backup2/member/snap/0000000000000002-000000000001fbdd.snap
backup2/member/wal
backup2/member/wal/0000000000000000-0000000000000000.wal
```

With these missing WAL files etcd crashes:

```
2016-05-15 00:40:22.233119 I | raft: 54b1d9a59f01 became follower at term 2
2016-05-15 00:40:22.233180 I | raft: newRaft 54b1d9a59f01 [peers: [8c76dbadd15001d1], term: 2, commit: 132960, applied: 130013, lastindex: 132960, lastterm: 2]
2016-05-15 00:40:22.233317 I | etcdserver: added member 8c76dbadd15001d1 [http://10.0.0.50:2380] to cluster 54b1d9a59f02 from store
2016-05-15 00:40:22.233378 I | etcdserver: set the cluster version to 2.3 from store
2016-05-15 00:40:22.233976 I | etcdserver: starting server... [version: 2.3.4, cluster version: 2.3]
2016-05-15 00:40:22.234643 E | etcdmain: failed to notify systemd for readiness: No socket
2016-05-15 00:40:22.234695 E | etcdmain: forgot to set Type=notify in systemd service file?
panic: runtime error: index out of range

goroutine 69 [running]:
panic(0xdc2620, 0xc820010090)
        /usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/coreos/etcd/raft.(*raft).maybeCommit(0xc820b2e9c0, 0xc8201385a0)
        /home/gyuho/go/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/raft/raft.go:382 +0x302
github.com/coreos/etcd/raft.(*raft).removeNode(0xc820b2e9c0, 0x8c76dbadd15001d1)
        /home/gyuho/go/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/raft/raft.go:842 +0x59
github.com/coreos/etcd/raft.(*node).run(0xc820b9e050, 0xc820b2e9c0)
        /home/gyuho/go/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/raft/node.go:330 +0xca4
created by github.com/coreos/etcd/raft.RestartNode
        /home/gyuho/go/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/raft/node.go:215 +0x2e4
```

If I manually copy in the WAL things start to work:

```
core@ip-10-0-0-50 ~ $ sudo cp /var/lib/etcd2/member/wal/0000000000000001-0000000000011eec.wal backup3/member/wal/
core@ip-10-0-0-50 ~ $ sudo chmod -R 777 backup3
core@ip-10-0-0-50 ~ $ sudo ./etcd-v2.3.4-linux-amd64/etcd --force-new-cluster --data-dir backup3/
```

--- a/raft/raft.go
+++ b/raft/raft.go
@@ -893,6 +893,12 @@ func (r *raft) addNode(id uint64) {
 func (r *raft) removeNode(id uint64) {
 	r.delProgress(id)
 	r.pendingConf = false
+
+	// do not try to commit or abort transferring if there is no nodes in the cluster.
+	if len(r.prs) == 0 {
+		return
+	}
+
 	// The quorum size is now smaller, so see if any pending entries can
 	// be committed.
 	if r.maybeCommit() {

--- a/raft/raft_test.go
+++ b/raft/raft_test.go
@@ -1782,6 +1782,13 @@ func TestRemoveNode(t *testing.T) {
 	if g := r.nodes(); !reflect.DeepEqual(g, w) {
 		t.Errorf("nodes = %v, want %v", g, w)
 	}
+
+	// remove all nodes from cluster
+	r.removeNode(1)
+	w = []uint64{}
+	if g := r.nodes(); !reflect.DeepEqual(g, w) {
+		t.Errorf("nodes = %v, want %v", g, w)
+	}
 }
 
 func TestPromotable(t *testing.T) {

